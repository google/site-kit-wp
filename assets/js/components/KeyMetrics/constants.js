/**
 * Key Metrics components - constants.
 *
 * Site Kit by Google, Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

export const KEY_METRICS_SETUP_CTA_WIDGET_SLUG = 'key-metrics-setup-cta-widget';

export const KEY_METRICS_SELECTION_PANEL_OPENED_KEY =
	'googlesitekit-key-metrics-selection-panel-opened';
export const KEY_METRICS_SELECTION_FORM = 'key-metrics-selection-form';
export const KEY_METRICS_SELECTED = 'key-metrics-selected';
export const EFFECTIVE_SELECTION = 'key-metrics-effective-selection';
export const UNSTAGED_SELECTION = 'key-metrics-unstaged-selection';
export const MIN_SELECTED_METRICS_COUNT = 2;
export const MAX_SELECTED_METRICS_COUNT = 4;
export const MAX_SELECTED_METRICS_COUNT_WITH_CONVERSION_EVENTS = 8;
export const KEY_METRICS_GROUP_CURRENT = {
	SLUG: 'current-selection',
	LABEL: __( 'Current selection', 'google-site-kit' ),
};
export const KEY_METRICS_GROUP_SUGGESTED = {
	SLUG: 'suggested',
	LABEL: __( 'Suggested', 'google-site-kit' ),
};
export const KEY_METRICS_GROUP_VISITORS = {
	SLUG: 'visitors',
	LABEL: __( 'Visitors', 'google-site-kit' ),
};
export const KEY_METRICS_GROUP_DRIVING_TRAFFIC = {
	SLUG: 'driving-traffic',
	LABEL: __( 'Driving traffic', 'google-site-kit' ),
};
export const KEY_METRICS_GROUP_GENERATING_LEADS = {
	SLUG: 'generating-leads',
	LABEL: __( 'Generating leads', 'google-site-kit' ),
};
export const KEY_METRICS_GROUP_SELLING_PRODUCTS = {
	SLUG: 'selling-products',
	LABEL: __( 'Selling products', 'google-site-kit' ),
};
export const KEY_METRICS_GROUP_CONTENT_PERFORMANCE = {
	SLUG: 'content-performance',
	LABEL: __( 'Content performance', 'google-site-kit' ),
};
