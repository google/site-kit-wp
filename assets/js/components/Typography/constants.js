/**
 * Site Kit by Google, Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const TYPE_DISPLAY = 'display';
export const TYPE_HEADLINE = 'headline';
export const TYPE_TITLE = 'title';
export const TYPE_BODY = 'body';
export const TYPE_LABEL = 'label';

export const VALID_TYPES = [
	TYPE_BODY,
	TYPE_DISPLAY,
	TYPE_HEADLINE,
	TYPE_LABEL,
	TYPE_TITLE,
];

export const SIZE_SMALL = 'small';
export const SIZE_MEDIUM = 'medium';
export const SIZE_LARGE = 'large';

export const VALID_SIZES = [ SIZE_SMALL, SIZE_MEDIUM, SIZE_LARGE ];
