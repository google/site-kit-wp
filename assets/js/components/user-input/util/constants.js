/**
 * User Input Answers.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

export const USER_INPUT_QUESTIONS_PURPOSE = 'purpose';
export const USER_INPUT_QUESTION_POST_FREQUENCY = 'postFrequency';
export const USER_INPUT_QUESTIONS_GOALS = 'goals';

export const USER_INPUT_QUESTIONS_LIST = [
	USER_INPUT_QUESTIONS_PURPOSE,
	USER_INPUT_QUESTION_POST_FREQUENCY,
	USER_INPUT_QUESTIONS_GOALS,
];

export const USER_INPUT_MAX_ANSWERS = {
	[ USER_INPUT_QUESTIONS_PURPOSE ]: 1,
	[ USER_INPUT_QUESTION_POST_FREQUENCY ]: 1,
	[ USER_INPUT_QUESTIONS_GOALS ]: 3,
};

export const USER_INPUT_CURRENTLY_EDITING_KEY =
	'googlesitekit-user-input-currently-editing';

export const FORM_USER_INPUT_QUESTION_NUMBER = 'user_input_question_number';
export const FORM_USER_INPUT_QUESTION_SNAPSHOT = 'user_input_question_snapshot';
export const USER_INPUT_LEGACY_SITE_PURPOSE_DISMISSED_ITEM_KEY =
	'user-input-legacy-site-purpose-dismissed-item';

/**
 * Gets available questions for user input settings.
 *
 * @since 1.139.0
 * @private
 *
 * @return {Object} Questions object.
 */
export function getUserInputQuestions() {
	return [
		{
			title: __(
				'What is the main purpose of this site?',
				'google-site-kit'
			),
			description: __(
				'Based on your answer, Site Kit will tailor the metrics you see on your dashboard to help you track progress towards your specific goals',
				'google-site-kit'
			),
		},
		{
			title: __(
				'How often do you create new content for this site?',
				'google-site-kit'
			),
			description: __(
				'Based on your answer, Site Kit will suggest new features for your dashboard related to content creation',
				'google-site-kit'
			),
		},
		{
			title: __(
				'What are your top goals for this site?',
				'google-site-kit'
			),
			description: __(
				'Based on your answers, Site Kit will tailor the metrics and recommendations you see on your dashboard to help you make progress in these areas',
				'google-site-kit'
			),
		},
	];
}

/**
 * Gets available answers for user input settings.
 *
 * @since 1.20.0
 * @private
 *
 * @return {Object} Answers object.
 */
export function getUserInputAnswers() {
	return {
		USER_INPUT_ANSWERS_PURPOSE: {
			sell_products_or_service: __(
				'Sell products or services',
				'google-site-kit'
			),
			sell_products: __( 'Sell products', 'google-site-kit' ),
			provide_services: __( 'Provide services', 'google-site-kit' ),
			monetize_content: __(
				'Monetize content (with ads or affiliate links)',
				'google-site-kit'
			),
			publish_blog: __( 'Publish a blog', 'google-site-kit' ),
			publish_news: __( 'Publish news content', 'google-site-kit' ),
			share_portfolio: __(
				'Share a business card or portfolio to represent me or my company online',
				'google-site-kit'
			),
			other: __( 'Other', 'google-site-kit' ),
		},
		USER_INPUT_ANSWERS_POST_FREQUENCY: {
			never: __( 'Never', 'google-site-kit' ),
			daily: __( 'Daily', 'google-site-kit' ),
			weekly: __( 'Weekly', 'google-site-kit' ),
			monthly: __( 'Monthly', 'google-site-kit' ),
			other: __( 'Other', 'google-site-kit' ),
		},
		USER_INPUT_ANSWERS_GOALS: {
			retaining_visitors: __(
				'Retain visitors, turn them into loyal readers or customers',
				'google-site-kit'
			),
			improving_performance: __(
				'Improve speed and performance',
				'google-site-kit'
			),
			finding_new_topics: __(
				'Find new topics to write about that connect with my audience',
				'google-site-kit'
			),
			growing_audience: __( 'Grow my audience', 'google-site-kit' ),
			expanding_business: __(
				'Expand my business into new cities, states or markets',
				'google-site-kit'
			),
			generating_revenue: __(
				'Generate more revenue',
				'google-site-kit'
			),
			help_better_rank: __(
				'Help my content rank in a better position in Google search results',
				'google-site-kit'
			),
			understanding_content_performance: __(
				'Understand which content is performing best',
				'google-site-kit'
			),
			encourage_to_post: __(
				'Tips for generating and posting engaging content updates',
				'google-site-kit'
			),
			other: __( 'Other', 'google-site-kit' ),
		},
	};
}

/**
 * Gets available answer descriptions for user input settings.
 *
 * @since 1.139.0
 * @private
 *
 * @return {Object} Answer descriptions object.
 */
export function getUserInputAnswersDescription() {
	return {
		USER_INPUT_ANSWERS_PURPOSE: {
			sell_products_or_service: __(
				'E.g. selling products like devices, apparel, equipment, etc. or offering services like courses, consulting, tutoring, etc.',
				'google-site-kit'
			),
			sell_products: __(
				'E.g. selling devices, apparel, equipment, etc.',
				'google-site-kit'
			),
			provide_services: __(
				'E.g. offering courses, consulting, tutoring, etc.',
				'google-site-kit'
			),
			monetize_content: __(
				'Using display ads, affiliate links, sponsored content, etc.',
				'google-site-kit'
			),
			publish_blog: __(
				'Writing on a topic youâ€™re passionate about, no focus on monetizing content',
				'google-site-kit'
			),
			publish_news: __(
				'E.g. local news, investigative pieces, interviews, etc.',
				'google-site-kit'
			),
			share_portfolio: __(
				'My website represents me or my company',
				'google-site-kit'
			),
			other: undefined,
		},
	};
}
