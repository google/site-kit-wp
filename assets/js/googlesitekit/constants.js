/**
 * Core constants.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export const VIEW_CONTEXT_MAIN_DASHBOARD = 'mainDashboard';
export const VIEW_CONTEXT_ENTITY_DASHBOARD = 'entityDashboard';
export const VIEW_CONTEXT_MAIN_DASHBOARD_VIEW_ONLY = 'mainDashboardViewOnly';
export const VIEW_CONTEXT_ENTITY_DASHBOARD_VIEW_ONLY =
	'entityDashboardViewOnly';
export const VIEW_CONTEXT_POSTS_LIST = 'postsList';
export const VIEW_CONTEXT_USER_INPUT = 'userInput';
export const VIEW_CONTEXT_ACTIVATION = 'activation';
export const VIEW_CONTEXT_SPLASH = 'splash';
export const VIEW_CONTEXT_ADMIN_BAR = 'adminBar';
export const VIEW_CONTEXT_ADMIN_BAR_VIEW_ONLY = 'adminBarViewOnly';
export const VIEW_CONTEXT_SETTINGS = 'settings';
export const VIEW_CONTEXT_AD_BLOCKING_RECOVERY = 'adBlockingRecovery';
export const VIEW_CONTEXT_MODULE = 'module';
export const VIEW_CONTEXT_WP_DASHBOARD = 'wpDashboard';
export const VIEW_CONTEXT_WP_DASHBOARD_VIEW_ONLY = 'wpDashboardViewOnly';
export const VIEW_CONTEXT_MODULE_SETUP = 'moduleSetup';
export const VIEW_CONTEXT_METRIC_SELECTION = 'metricSelection';

// Anchor ID Constants
export const ANCHOR_ID_KEY_METRICS = 'key-metrics';
export const ANCHOR_ID_TRAFFIC = 'traffic';
export const ANCHOR_ID_CONTENT = 'content';
export const ANCHOR_ID_SPEED = 'speed';
export const ANCHOR_ID_MONETIZATION = 'monetization';

// Site Kit only view contexts
export const SITE_KIT_VIEW_CONTEXTS = [
	VIEW_CONTEXT_MAIN_DASHBOARD,
	VIEW_CONTEXT_ENTITY_DASHBOARD,
	VIEW_CONTEXT_MAIN_DASHBOARD_VIEW_ONLY,
	VIEW_CONTEXT_ENTITY_DASHBOARD_VIEW_ONLY,
	VIEW_CONTEXT_USER_INPUT,
	VIEW_CONTEXT_SPLASH,
	VIEW_CONTEXT_SETTINGS,
	VIEW_CONTEXT_MODULE_SETUP,
	VIEW_CONTEXT_METRIC_SELECTION,
];

// Site Kit view-only contexts
export const SITE_KIT_VIEW_ONLY_CONTEXTS = [
	VIEW_CONTEXT_MAIN_DASHBOARD_VIEW_ONLY,
	VIEW_CONTEXT_ENTITY_DASHBOARD_VIEW_ONLY,
	VIEW_CONTEXT_ADMIN_BAR_VIEW_ONLY,
	VIEW_CONTEXT_WP_DASHBOARD_VIEW_ONLY,
];
