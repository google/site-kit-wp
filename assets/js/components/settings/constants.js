/**
 * Settings constants.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { MODULE_SLUG_ADS } from '../../modules/ads/datastore/constants';
import { MODULE_SLUG_SIGN_IN_WITH_GOOGLE } from '../../modules/sign-in-with-google/datastore/constants';
import { MODULE_SLUG_READER_REVENUE_MANAGER } from '../../modules/reader-revenue-manager/datastore/constants';

export const NEW_MODULES = [
	MODULE_SLUG_ADS,
	MODULE_SLUG_READER_REVENUE_MANAGER,
	MODULE_SLUG_SIGN_IN_WITH_GOOGLE,
];

export const BETA_MODULES = [ MODULE_SLUG_SIGN_IN_WITH_GOOGLE ];

export const EXPERIMENTAL_MODULES = [];
