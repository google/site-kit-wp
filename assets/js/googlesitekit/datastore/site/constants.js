/**
 * `core/site` data store: constants.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { MODULE_SLUG_ADS } from '@/js/modules/ads/constants';
import { MODULE_SLUG_ANALYTICS_4 } from '@/js/modules/analytics-4/constants';
import { MODULE_SLUG_TAGMANAGER } from '@/js/modules/tagmanager/constants';

export const CORE_SITE = 'core/site';

export const AMP_MODE_PRIMARY = 'primary';
export const AMP_MODE_SECONDARY = 'secondary';

export const GOOGLE_TAG_GATEWAY_MODULES = [
	MODULE_SLUG_ANALYTICS_4,
	MODULE_SLUG_ADS,
	MODULE_SLUG_TAGMANAGER,
];
