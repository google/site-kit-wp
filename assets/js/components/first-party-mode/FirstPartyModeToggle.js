/**
 * First Party Mode Toggle component.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import classnames from 'classnames';
import { useMount } from 'react-use';

/**
 * WordPress dependencies
 */
import { createInterpolateElement, useCallback } from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { ProgressBar, Switch } from 'googlesitekit-components';
import { useDispatch, useSelect } from 'googlesitekit-data';
import { CORE_SITE } from '../../googlesitekit/datastore/site/constants';
import Badge from '../Badge';
import SubtleNotification from '../notifications/SubtleNotification';
import Link from '../Link';

export default function FirstPartyModeToggle( { className } ) {
	const isFirstPartyModeEnabled = useSelect( ( select ) =>
		select( CORE_SITE ).isFirstPartyModeEnabled()
	);
	const isLoading = useSelect( ( select ) =>
		select( CORE_SITE ).isFetchingGetFPMServerRequirementStatus()
	);
	const hasMetServerRequirements = useSelect( ( select ) => {
		const { isFPMHealthy, isScriptAccessEnabled } = select( CORE_SITE );

		return isFPMHealthy() !== false && isScriptAccessEnabled() !== false;
	} );

	const { fetchGetFPMServerRequirementStatus, setFirstPartyModeEnabled } =
		useDispatch( CORE_SITE );

	const learnMoreURL = useSelect( ( select ) => {
		return select( CORE_SITE ).getDocumentationLinkURL(
			'first-party-mode-introduction'
		);
	} );

	const serverRequirementsLearnMoreURL = useSelect( ( select ) => {
		return select( CORE_SITE ).getDocumentationLinkURL(
			'first-party-mode-server-requirements'
		);
	} );

	// Fetch the server requirement status on mount.
	useMount( fetchGetFPMServerRequirementStatus );

	const handleClick = useCallback( () => {
		setFirstPartyModeEnabled( ! isFirstPartyModeEnabled );
	}, [ isFirstPartyModeEnabled, setFirstPartyModeEnabled ] );

	return (
		<div
			className={ classnames(
				'googlesitekit-first-party-mode-toggle',
				className
			) }
		>
			{ isLoading && (
				<ProgressBar
					small
					className="googlesitekit-first-party-mode-toggle__progress"
				/>
			) }
			{ ! isLoading && (
				<div className="googlesitekit-module-settings-group__switch">
					<Switch
						label={ __( 'First-party mode', 'google-site-kit' ) }
						checked={
							!! isFirstPartyModeEnabled &&
							hasMetServerRequirements
						}
						disabled={ ! hasMetServerRequirements }
						onClick={ handleClick }
						hideLabel={ false }
					/>
					<div className="googlesitekit-first-party-mode-toggle__switch-badge">
						<Badge
							className="googlesitekit-badge--beta"
							hasLeftSpacing
							label={ __( 'Beta', 'google-site-kit' ) }
						/>
					</div>
				</div>
			) }
			<p className="googlesitekit-module-settings-group__helper-text">
				{ createInterpolateElement(
					__(
						'Your tag data will be sent through your own domain to improve data quality and help you recover measurement signals. <a>Learn more</a>',
						'google-site-kit'
					),
					{
						a: (
							<Link
								href={ learnMoreURL }
								external
								aria-label={ __(
									'Learn more about first-party mode',
									'google-site-kit'
								) }
							/>
						),
					}
				) }
			</p>
			{ ! isLoading && ! hasMetServerRequirements && (
				<SubtleNotification
					title={ createInterpolateElement(
						__(
							'Your serverâ€™s current settings prevent first-party mode from working. To enable it, please contact your hosting provider and request access to external resources and plugin files. <a>Learn more</a>',
							'google-site-kit'
						),
						{
							a: (
								<Link
									href={ serverRequirementsLearnMoreURL }
									external
									aria-label={ __(
										'Learn more about first-party mode server requirements',
										'google-site-kit'
									) }
								/>
							),
						}
					) }
					variant="warning"
				/>
			) }
		</div>
	);
}
