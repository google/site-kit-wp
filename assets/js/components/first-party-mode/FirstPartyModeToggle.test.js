/**
 * First Party Mode Toggle component tests.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import {
	render,
	createTestRegistry,
	muteFetch,
	freezeFetch,
} from '../../../../tests/js/test-utils';
import { CORE_SITE } from '../../googlesitekit/datastore/site/constants';
import FirstPartyModeToggle from './FirstPartyModeToggle';

describe( 'FirstPartyModeToggle', () => {
	let registry;

	const serverRequirementStatusEndpoint = new RegExp(
		'^/google-site-kit/v1/core/site/data/fpm-server-requirement-status'
	);

	beforeEach( () => {
		registry = createTestRegistry();

		registry.dispatch( CORE_SITE ).receiveGetFirstPartyModeSettings( {
			isEnabled: false,
			isFPMHealthy: null,
			isScriptAccessEnabled: null,
		} );
	} );

	it( 'should make a request to fetch the server requirement status', async () => {
		muteFetch( serverRequirementStatusEndpoint );

		const { waitForRegistry } = render( <FirstPartyModeToggle />, {
			registry,
		} );

		await waitForRegistry();

		expect( fetchMock ).toHaveFetched( serverRequirementStatusEndpoint );
	} );

	it( 'should render in loading state if the server requirement status is still loading', async () => {
		freezeFetch( serverRequirementStatusEndpoint );

		const { container, getByRole, waitForRegistry } = render(
			<FirstPartyModeToggle />,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect( getByRole( 'progressbar' ) ).toBeInTheDocument();

		expect( container ).toMatchSnapshot();
	} );

	it( 'should render in default state', async () => {
		fetchMock.getOnce( serverRequirementStatusEndpoint, {
			body: {
				isEnabled: false,
				isFPMHealthy: true,
				isScriptAccessEnabled: true,
			},
			status: 200,
		} );

		const { container, getByLabelText, waitForRegistry } = render(
			<FirstPartyModeToggle />,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect( getByLabelText( 'First-party mode' ) ).toBeInTheDocument();

		expect( container ).toMatchSnapshot();
	} );

	it( 'should render in disabled state if server requirements are not met', async () => {
		fetchMock.getOnce( serverRequirementStatusEndpoint, {
			body: {
				isEnabled: false,
				isFPMHealthy: false,
				isScriptAccessEnabled: false,
			},
			status: 200,
		} );

		const { container, getByLabelText, waitForRegistry } = render(
			<FirstPartyModeToggle />,
			{
				registry,
			}
		);

		await waitForRegistry();

		expect( getByLabelText( 'First-party mode' ) ).toBeDisabled();

		expect( container ).toHaveTextContent(
			'Your serverâ€™s current settings prevent first-party mode from working. To enable it, please contact your hosting provider and request access to external resources and plugin files.'
		);

		expect( container ).toMatchSnapshot();
	} );

	it.each( [ 'isFPMHealthy', 'isScriptAccessEnabled' ] )(
		'should not render in disabled state unless %s is explicitly false',
		async ( requirement ) => {
			const response = {
				isEnabled: false,
				isFPMHealthy: true,
				isScriptAccessEnabled: true,
			};

			response[ requirement ] = null;

			fetchMock.getOnce( serverRequirementStatusEndpoint, {
				body: response,
				status: 200,
			} );

			const { getByLabelText, waitForRegistry } = render(
				<FirstPartyModeToggle />,
				{
					registry,
				}
			);

			await waitForRegistry();

			expect( getByLabelText( 'First-party mode' ) ).not.toBeDisabled();
		}
	);

	it.each( [ 'isFPMHealthy', 'isScriptAccessEnabled' ] )(
		'should render in disabled state if %s is false',
		async ( requirement ) => {
			const response = {
				isEnabled: false,
				isFPMHealthy: true,
				isScriptAccessEnabled: true,
			};

			response[ requirement ] = false;

			fetchMock.getOnce( serverRequirementStatusEndpoint, {
				body: response,
				status: 200,
			} );

			const { getByLabelText, waitForRegistry } = render(
				<FirstPartyModeToggle />,
				{
					registry,
				}
			);

			await waitForRegistry();

			expect( getByLabelText( 'First-party mode' ) ).toBeDisabled();
		}
	);

	it( 'should toggle first party mode on click', async () => {
		fetchMock.getOnce( serverRequirementStatusEndpoint, {
			body: {
				isEnabled: false,
				isFPMHealthy: true,
				isScriptAccessEnabled: true,
			},
			status: 200,
		} );

		const { getByLabelText, waitForRegistry } = render(
			<FirstPartyModeToggle />,
			{
				registry,
			}
		);

		await waitForRegistry();

		const switchControl = getByLabelText( 'First-party mode' );

		expect( switchControl ).not.toBeChecked();

		expect( registry.select( CORE_SITE ).isFirstPartyModeEnabled() ).toBe(
			false
		);

		switchControl.click();

		expect( switchControl ).toBeChecked();

		expect( registry.select( CORE_SITE ).isFirstPartyModeEnabled() ).toBe(
			true
		);

		// Give it another click to verify it can be toggled off.
		switchControl.click();

		expect( switchControl ).not.toBeChecked();

		expect( registry.select( CORE_SITE ).isFirstPartyModeEnabled() ).toBe(
			false
		);
	} );

	it( 'should render a "Beta" badge', async () => {
		fetchMock.getOnce( serverRequirementStatusEndpoint, {
			body: {
				isEnabled: false,
				isFPMHealthy: true,
				isScriptAccessEnabled: true,
			},
			status: 200,
		} );

		const { container, waitForRegistry } = render(
			<FirstPartyModeToggle />,
			{
				registry,
			}
		);

		await waitForRegistry();

		const badgeElement = container.querySelector( '.googlesitekit-badge' );

		expect( badgeElement ).toBeInTheDocument();
		expect( badgeElement ).toHaveTextContent( 'Beta' );
	} );
} );
