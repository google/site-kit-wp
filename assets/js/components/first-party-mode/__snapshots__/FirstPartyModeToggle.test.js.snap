// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FirstPartyModeToggle should render in default state 1`] = `
<div>
  <div
    class="googlesitekit-first-party-mode-toggle"
  >
    <div
      class="googlesitekit-module-settings-group__switch"
    >
      <div
        aria-checked="false"
        class="mdc-switch"
        role="switch"
        tabindex="0"
      >
        <div
          class="mdc-switch__track"
        >
           
        </div>
        <div
          class="mdc-switch__thumb-underlay"
        >
          <div
            class="mdc-switch__thumb"
          >
            <input
              class="mdc-switch__native-control"
              id="googlesitekit-switch-4"
              readonly=""
              role="switch"
              type="checkbox"
            />
          </div>
        </div>
      </div>
      <label
        class=""
        for="googlesitekit-switch-4"
      >
        First-party mode
      </label>
      <div
        class="googlesitekit-first-party-mode-toggle__switch-badge"
      >
        <span
          class="googlesitekit-badge googlesitekit-badge--beta googlesitekit-badge--has-left-spacing"
        >
          Beta
        </span>
      </div>
    </div>
    <p
      class="googlesitekit-module-settings-group__helper-text"
    >
      Your tag data will be sent through your own domain to improve data quality and help you recover measurement signals. 
      <a
        aria-label="Learn more about first-party mode (opens in a new tab)"
        class="googlesitekit-cta-link"
        href="undefined?doc=first-party-mode-introduction"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span
          class="googlesitekit-cta-link__contents"
        >
          Learn more
        </span>
        <span
          class="googlesitekit-icon-wrapper"
          style="margin-left: 5px;"
        >
          <svg />
        </span>
      </a>
    </p>
  </div>
</div>
`;

exports[`FirstPartyModeToggle should render in disabled state if server requirements are not met 1`] = `
<div>
  <div
    class="googlesitekit-first-party-mode-toggle"
  >
    <div
      class="googlesitekit-module-settings-group__switch"
    >
      <div
        aria-checked="false"
        class="mdc-switch mdc-switch--disabled"
        role="switch"
        tabindex="0"
      >
        <div
          class="mdc-switch__track"
        >
           
        </div>
        <div
          class="mdc-switch__thumb-underlay"
        >
          <div
            class="mdc-switch__thumb"
          >
            <input
              class="mdc-switch__native-control"
              disabled=""
              id="googlesitekit-switch-6"
              readonly=""
              role="switch"
              type="checkbox"
            />
          </div>
        </div>
      </div>
      <label
        class=""
        for="googlesitekit-switch-6"
      >
        First-party mode
      </label>
      <div
        class="googlesitekit-first-party-mode-toggle__switch-badge"
      >
        <span
          class="googlesitekit-badge googlesitekit-badge--beta googlesitekit-badge--has-left-spacing"
        >
          Beta
        </span>
      </div>
    </div>
    <p
      class="googlesitekit-module-settings-group__helper-text"
    >
      Your tag data will be sent through your own domain to improve data quality and help you recover measurement signals. 
      <a
        aria-label="Learn more about first-party mode (opens in a new tab)"
        class="googlesitekit-cta-link"
        href="undefined?doc=first-party-mode-introduction"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span
          class="googlesitekit-cta-link__contents"
        >
          Learn more
        </span>
        <span
          class="googlesitekit-icon-wrapper"
          style="margin-left: 5px;"
        >
          <svg />
        </span>
      </a>
    </p>
    <div
      class="googlesitekit-subtle-notification googlesitekit-subtle-notification--warning"
    >
      <div
        class="googlesitekit-subtle-notification__icon"
      >
        <svg />
      </div>
      <div
        class="googlesitekit-subtle-notification__content"
      >
        <p>
          Your server’s current settings prevent first-party mode from working. To enable it, please contact your hosting provider and request access to external resources and plugin files. 
          <a
            aria-label="Learn more about first-party mode server requirements (opens in a new tab)"
            class="googlesitekit-cta-link"
            href="undefined?doc=first-party-mode-server-requirements"
            rel="noopener noreferrer"
            target="_blank"
          >
            <span
              class="googlesitekit-cta-link__contents"
            >
              Learn more
            </span>
            <span
              class="googlesitekit-icon-wrapper"
              style="margin-left: 5px;"
            >
              <svg />
            </span>
          </a>
        </p>
      </div>
      <div
        class="googlesitekit-subtle-notification__action"
      />
    </div>
  </div>
</div>
`;

exports[`FirstPartyModeToggle should render in loading state if the server requirement status is still loading 1`] = `
<div>
  <div
    class="googlesitekit-first-party-mode-toggle"
  >
    <div
      class="mdc-linear-progress googlesitekit-first-party-mode-toggle__progress mdc-linear-progress--indeterminate mdc-linear-progress--small"
      role="progressbar"
    >
      <div
        class="mdc-linear-progress__buffering-dots"
      />
      <div
        class="mdc-linear-progress__buffer"
      />
      <div
        class="mdc-linear-progress__bar mdc-linear-progress__primary-bar"
      >
        <span
          class="mdc-linear-progress__bar-inner"
        />
      </div>
      <div
        class="mdc-linear-progress__bar mdc-linear-progress__secondary-bar"
      >
        <span
          class="mdc-linear-progress__bar-inner"
        />
      </div>
    </div>
    <p
      class="googlesitekit-module-settings-group__helper-text"
    >
      Your tag data will be sent through your own domain to improve data quality and help you recover measurement signals. 
      <a
        aria-label="Learn more about first-party mode (opens in a new tab)"
        class="googlesitekit-cta-link"
        href="undefined?doc=first-party-mode-introduction"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span
          class="googlesitekit-cta-link__contents"
        >
          Learn more
        </span>
        <span
          class="googlesitekit-icon-wrapper"
          style="margin-left: 5px;"
        >
          <svg />
        </span>
      </a>
    </p>
  </div>
</div>
`;
